### PR TITLE
pss updates + kademlia fixes

### PR DESCRIPTION
**Pull request by lmars**

### CODE REVIEW

File: swarm/network/simulations/discovery/discovery_test.go [hunk 1]
```diff
@@ -135,6 +146,22 @@ func testDiscoverySimulation(t *testing.T, adapter adapters.NodeAdapter) {
...
+		snap, err := net.Snapshot()
+		if err != nil {
+			t.Fatalf("no shapshot dude")
+		}
+		jsonsnapshot, err := json.Marshal(snap)
+		if err != nil {
+			t.Fatalf("corrupt json snapshot: %v", err)
+		}
+		log.Info("writing snapshot", "file", *snapshotFile)
+		err = ioutil.WriteFile(*snapshotFile, jsonsnapshot, os.ModePerm)
```
Comment (lmars) [inline swarm/network/simulations/discovery/discovery_test.go]: Probably worth giving the file reasonable permissions like `0755`

File: swarm/swarm.go [hunk 2]
```diff
@@ -1,3 +1,5 @@
+// +build !psschat
```
Comment (lmars) [inline swarm/swarm.go]: I don't think references to `psschat` belong in the swarm codebase.

File: swarm/swarm.go [hunk 3]
```diff
@@ -38,6 +42,7 @@ import (
 	"github.com/ethereum/go-ethereum/swarm/network"
 	"github.com/ethereum/go-ethereum/swarm/pss"
 	"github.com/ethereum/go-ethereum/swarm/storage"
+	//psschat "github.com/ethereum/go-ethereum/swarm/pss/protocols/chat"
```
Comment (lmars) [inline swarm/swarm.go]: Please remove.

File: swarm/swarm.go [hunk 4]
```diff
@@ -228,9 +253,9 @@ func (self *Swarm) Stop() error {
 	self.dpa.Stop()
 	//self.hive.Stop()
 	self.bzz.Stop()
-	if self.pss != nil {
-		self.pss.Stop()
-	}
+	//if self.pss != nil {
+	//	self.pss.Stop()
+	//}
```
Comment (lmars) [inline swarm/swarm.go]: Why is this commented out?

File: swarm/swarm_psschat.go [hunk 5]
```diff
@@ -0,0 +1,415 @@
+// +build psschat
```
Comment (lmars) [inline swarm/swarm_psschat.go]: This file should not be part of swarm.

**Comment by lmars** [comment]
@nolash I have rebased your further changes, looks good but I got the following panic when running the tests:

```
fatal error: concurrent map writes

goroutine 2313 [running]:
runtime.throw(0x471e968, 0x15)
        /usr/local/Cellar/go/1.8.1/libexec/src/runtime/panic.go:596 +0x95 fp=0xc420653530 sp=0xc420653510
runtime.mapassign(0x4648600, 0xc42cbd8a20, 0xc420653630, 0xc42d07d108)
        /usr/local/Cellar/go/1.8.1/libexec/src/runtime/hashmap.go:499 +0x667 fp=0xc4206535d0 sp=0xc420653530
github.com/ethereum/go-ethereum/swarm/pss.(*Pss).Run(0xc42cbc8be0, 0xc42c70c840, 0x4a2efe0, 0xc42f6c6310, 0xc426204340, 0xc420653758)
        /Users/lewis/go/src/github.com/ethereum/go-ethereum/swarm/pss/pss.go:137 +0x1c8 fp=0xc4206536a8 sp=0xc4206535d0
github.com/ethereum/go-ethereum/swarm/pss.(*Pss).Run-fm(0xc42c70c840, 0x4a2efe0, 0xc42f6c6310, 0x0, 0x0)
        /Users/lewis/go/src/github.com/ethereum/go-ethereum/swarm/pss/pss.go:127 +0x48 fp=0xc4206536e8 sp=0xc4206536a8
github.com/ethereum/go-ethereum/p2p.(*Peer).startProtocols.func1(0xc42f6c6310, 0xc42c70c840, 0x4a2efe0, 0xc42f6c6310)
        /Users/lewis/go/src/github.com/ethereum/go-ethereum/p2p/peer.go:340 +0x66 fp=0xc4206537c0 sp=0xc4206536e8
runtime.goexit()
        /usr/local/Cellar/go/1.8.1/libexec/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4206537c8 sp=0xc4206537c0
created by github.com/ethereum/go-ethereum/p2p.(*Peer).startProtocols
        /Users/lewis/go/src/github.com/ethereum/go-ethereum/p2p/peer.go:349 +0x2be
```

Looks like we need protection of the [fwdPool field](https://github.com/ethersphere/go-ethereum/blob/network-testing-framework-psstalk/swarm/pss/pss.go#L137).

**Comment by lmars** [comment]
@zelig can you respond to the pot related questions?

File: swarm/pss/client/client.go [hunk 6]
```diff
@@ -20,44 +19,32 @@ import (
...
+	// rpc connections
+	rpc *rpc.Client
+	sub *rpc.ClientSubscription
+
+	// channels
+	topicsC chan []byte
+	msgC    chan pss.APIMsg
+	quitC   chan struct{}
+
+	ctx    context.Context
```
Comment (lmars) [inline swarm/pss/client/client.go]: The context should not be a struct field, all methods which require a context should accept it as a first argument, usually named `ctx`.

The main reason is that if I pass a context with a timeout, I want that timeout to only affect the immediate method I am calling, not affect all future calls on an object.

See https://golang.org/pkg/context/#pkg-overview for reference.

File: swarm/pss/client/client.go [hunk 7]
```diff
@@ -17,6 +18,7 @@ import (
 )
 
 const (
+	rpcTimeout     = time.Second
```
Comment (lmars) [inline swarm/pss/client/client.go]: I we have a default timeout, we should give users the ability to change that timeout, but in my opinion there should not even be a default timeout. If the user needs a timeout, they should pass in a context which has the timeout they want to use.

Comment (nolash) [reply swarm/pss/client/client.go]: so you mean if context that get passed is nil, we use the background?

Comment (lmars) [reply swarm/pss/client/client.go]: [As per the docs](https://golang.org/pkg/context/#pkg-overview), users are discouraged from passing a nil context so we should assume it is always set.

**Pull request merged**